flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaWriter.java

```diff
@@ -0,0 +1,379 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.connectors.kafka.sink;
+
+import org.apache.flink.api.common.serialization.SerializationSchema;
+import org.apache.flink.api.connector.sink.Sink;
+import org.apache.flink.api.connector.sink.SinkWriter;
+import org.apache.flink.connector.base.DeliveryGuarantee;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.streaming.connectors.kafka.internals.metrics.KafkaMetricMutableWrapper;
+
+import org.apache.flink.shaded.guava30.com.google.common.collect.ImmutableList;
+
+import org.apache.kafka.clients.producer.Callback;
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.Metric;
+import org.apache.kafka.common.MetricName;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * This class is responsible to write records in a Kafka topic and to handle the different delivery
+ * {@link DeliveryGuarantee}s.
+ *
+ * @param <IN> The type of the input elements.
+ */
+class KafkaWriter<IN> implements SinkWriter<IN, KafkaCommittable, KafkaWriterState> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KafkaWriter.class);
+    private static final String KEY_DISABLE_METRICS = "flink.disable-metrics";
+    private static final String KAFKA_PRODUCER_METRIC_NAME = "KafkaProducer";
+
+    private final DeliveryGuarantee deliveryGuarantee;
+    private final Properties kafkaProducerConfig;
+    private final String transactionalIdPrefix;
+    private final KafkaRecordSerializationSchema<IN> recordSerializer;
+    private final Callback deliveryCallback;
+    private final AtomicLong pendingRecords = new AtomicLong();
+    private final KafkaRecordSerializationSchema.KafkaSinkContext kafkaSinkContext;
+    private final List<FlinkKafkaInternalProducer<byte[], byte[]>> producers = new ArrayList<>();
+    private final Map<String, KafkaMetricMutableWrapper> previouslyCreatedMetrics = new HashMap<>();
+    private final MetricGroup metricGroup;
+
+    private transient FlinkKafkaInternalProducer<byte[], byte[]> currentProducer;
+    private transient KafkaWriterState kafkaWriterState;
```

Comment:
You should have a list of states in case of downscale (first entry should be the `main` or `active` entry).
Currently, we abort all transactions that are beyond this state, so if a restart happens before we committed everything, we may actually abort the transaction because we only store the `main` state in a new checkpoint. However, that would also not work with our enumeration of unknown task approach.

We probably need to encode the attempt after all ;) Can you create a follow-up ticket? The current limitation is that we assume that all transactions can be committed in-between downscale and the next checkpoint, which is okayish for a start.

Reply:
I think we already had this discussion ;) Let's clarify it offline I am pretty sure the current solution also covers your described case.